### DISCUSSION TITLE
Sub-second duration on entities card

### DISCUSSION THREAD

**Discussion by jhansche** [home-assistant/frontend#20893] [Dashboard] [answered]
This came out of an issue with how the new `sensor` entities work in the `ping` integration, and how it "feels" wrong compared to the old way of getting the same value as a Template sensor: https://github.com/home-assistant/core/issues/118183#issuecomment-2132367630

The short version is that the Template approach would render as `Router Ping ... 12.345 ms` (with units and sub-ms precision) in the entities card; while the new sensor renders as `Router ping ... 0.012` (no units, and the value is truncated at 1/1000 sec, or rounded to the nearest millisecond). Example from the link above:
![Screenshot 2024-05-26 at 2 59 42 PM](https://github.com/home-assistant/core/assets/99272/c4f5748b-6594-4eb1-8434-e21244db0536)

The reason is because the new Ping integration's sensor type declares `device_class: duration`, which is accurate. But the frontend renders a duration here with the assumption that it will be formatted as `"HH:mm:ss.SSS"`. For a Duration that is less than one second, that results in `ss.SSS` format, which is why the above screenshot only shows `0.001`.

Instead, I think it would make more sense to render these values with rules similar to what we see in [format_duration](https://github.com/home-assistant/frontend/blob/master/src/common/datetime/format_duration.ts#L7).

**Comment by karwosts** [comment, marked as answer]
If you set a `display_precision` on your entity, I believe it will not use HH:mm:ss format and instead render with the exact unit_of_measurement and requested number of decimal points.

**Reply by jhansche** [reply]
You're correct! I assumed that since it showed "Default (1.234)", that changing it would not make a difference. Setting it to "1.23" does work like you described. Nothing to do here, thanks!

**Reply by jhansche** [reply]
Related: is this something that the integration can choose to set in its `SensorEntityDescription` to work this way by default? 🤔

**Reply by karwosts** [reply]
I'm not sure, to be honest this is a bit of a hack/compromise on how to show duration sensors, it wasn't initially planned this way but display_precision wasn't otherwise used for duration sensors, so this was just added as a way to bypass the default display. 

Could probably benefit from some additional design on how to handle durations at various extremes, like multiple days or sub seconds.

**Reply by jhansche** [reply]
Looks like there is a [`suggested_display_precision`](https://github.com/home-assistant/core/blob/c34731185164aaf44419977c4086e9a7dd6c0a7f/homeassistant/components/sensor/__init__.py#L151). I'll propose that as an alternative for this scenario, given the `ping` integration's round-trip time will virtually never exceed the order of "seconds".

**Reply by karwosts** [reply]
Looking at frontend code I think that might not work, I think it has to be the user selected precision.

If you (or the ping integration author) wants to try that route, and you find it's not working, open a frontend issue and I'll try to look further.

**Reply by karwosts** [reply]
Hmm yeah history stats already seems to set suggested_display_precision = 2, for a duration sensor, and that still shows as hh:mm:ss.
